ext/pgsql/pgsql.c

```diff
@@ -3211,18 +3289,19 @@ PHP_FUNCTION(pg_copy_from)
 					if (UNEXPECTED(!tmp)) {
 						return;
 					}
-					query = (char *)emalloc(ZSTR_LEN(tmp) + 2);
-					strlcpy(query, ZSTR_VAL(tmp), ZSTR_LEN(tmp) + 2);
-					if (ZSTR_LEN(tmp) > 0 && *(query + ZSTR_LEN(tmp) - 1) != '\n') {
-						strlcat(query, "\n", ZSTR_LEN(tmp) + 2);
+					zend_string *zquery = zend_string_alloc(ZSTR_LEN(tmp) + 2, false);
+					memcpy(ZSTR_VAL(zquery), ZSTR_VAL(tmp), ZSTR_LEN(tmp) + 1);
+					ZSTR_VAL(zquery)[ZSTR_LEN(tmp) + 1] = 0;
+					if (ZSTR_LEN(tmp) > 0 && ZSTR_VAL(zquery)[ZSTR_LEN(tmp)]  != '\n') {
+						ZSTR_VAL(zquery)[ZSTR_LEN(tmp)] = '\n';
 					}
-					if (PQputCopyData(pgsql, query, (int)strlen(query)) != 1) {
-						efree(query);
+					if (PQputCopyData(pgsql, ZSTR_VAL(zquery), ZSTR_LEN(tmp) > 0 ? ZSTR_LEN(tmp) + 1 : 0) != 1) {
```

Comment:
I see what you do here, and I'm not sure it matters that we can send over an extra NUL terminator.
However, wouldn't it be better to simply adjust the length of the string `zquery` if it didn't have \n at the end initially?
I also realize now that you only need one additional byte instead of 2 for zend_string_alloc.